api/backend/ecs/environment_manager.go

```diff
@@ -327,6 +328,102 @@ func (this *ECSEnvironmentManager) DeleteEnvironment(environmentID string) error
 	return nil
 }
 
+func (this *ECSEnvironmentManager) CreateEnvironmentLink(sourceEnvironmentID, destEnvironmentID string) error {
+	sourceECSID := id.L0EnvironmentID(sourceEnvironmentID).ECSEnvironmentID()
+	destECSID := id.L0EnvironmentID(destEnvironmentID).ECSEnvironmentID()
+
+	sourceGroup, err := this.getEnvironmentSecurityGroup(sourceECSID)
+	if err != nil {
+		return err
+	}
+
+	destGroup, err := this.getEnvironmentSecurityGroup(destECSID)
+	if err != nil {
+		return err
+	}
+
+	if err := this.EC2.AuthorizeSecurityGroupIngressFromGroup(sourceGroup.GroupId, destGroup.GroupId); err != nil {
+		if !ContainsErrCode(err, "InvalidPermission.Duplicate") {
+			return err
+		}
+	}
+
+	if err := this.EC2.AuthorizeSecurityGroupIngressFromGroup(destGroup.GroupId, sourceGroup.GroupId); err != nil {
+		if !ContainsErrCode(err, "InvalidPermission.Duplicate") {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (this *ECSEnvironmentManager) DeleteEnvironmentLink(sourceEnvironmentID, destEnvironmentID string) error {
+	sourceECSID := id.L0EnvironmentID(sourceEnvironmentID).ECSEnvironmentID()
+	destECSID := id.L0EnvironmentID(destEnvironmentID).ECSEnvironmentID()
+
+	sourceGroup, err := this.EC2.DescribeSecurityGroup(sourceECSID.SecurityGroupName())
+	if err != nil {
+		return err
+	}
+
+	if sourceGroup == nil {
+		log.Warnf("Skipping environment unlink since security group '%s' does not exist", sourceECSID.SecurityGroupName())
+		return nil
+	}
+
+	destGroup, err := this.EC2.DescribeSecurityGroup(destECSID.SecurityGroupName())
+	if err != nil {
+		return err
+	}
+
+	if destGroup == nil {
+		log.Warnf("Skipping environment unlink since security group '%s' does not exist", destECSID.SecurityGroupName())
+		return nil
+	}
+
+	removeIngressRule := func(group *ec2.SecurityGroup, groupIDToRemove string) error {
+		for _, permission := range group.IpPermissions {
+			for _, pair := range permission.UserIdGroupPairs {
+				if *pair.GroupId == groupIDToRemove {
+					groupPermission := &awsec2.IpPermission{
+						IpProtocol:       permission.IpProtocol,
+						UserIdGroupPairs: []*awsec2.UserIdGroupPair{pair},
+					}
+
+					if err := this.EC2.RevokeSecurityGroupIngressHelper(*group.GroupId, groupPermission); err != nil {
+						return err
+					}
+				}
+			}
+		}
+
+		return nil
+	}
+
+	if err := removeIngressRule(sourceGroup, *destGroup.GroupId); err != nil {
+		return err
+	}
+
+	if err := removeIngressRule(destGroup, *sourceGroup.GroupId); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (this *ECSEnvironmentManager) getEnvironmentSecurityGroup(environmentID id.ECSEnvironmentID) (*ec2.SecurityGroup, error) {
+	group, err := this.EC2.DescribeSecurityGroup(environmentID.SecurityGroupName())
+	if err != nil {
+		return nil, err
+	}
+
+	if group == nil {
+		return nil, fmt.Errorf("Security group for environment '%s' does not exist", environmentID.L0EnvironmentID())
```

Comment:
this is a helper function used only in creating the link. It obviously can't create a link if the security group doesn't exist. However, delete actions need to be idempotent. If the security group doesn't exist, no unlinking work needs to be done, which satisfies the idempotency.